File: app/src/main/java/in/kashewdevelopers/todo/adapters/TaskRecycleListAdapter.java
package in.kashewdevelopers.todo.adapters;

import android.content.Context;
import android.database.Cursor;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;


import in.kashewdevelopers.todo.R;
import in.kashewdevelopers.todo.data_container.TaskDataContainer;

public class TaskRecycleListAdapter extends RecyclerView.Adapter<TaskViewHolder> {

    private Cursor cursor;
    private Context context;

    public TaskRecycleListAdapter(Context context, Cursor cursor) {
        this.cursor = cursor;
        this.context = context;
    }


    @NonNull
    @Override
    public TaskViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.task_list_item_layout, parent, false);

        return new TaskViewHolder(view)
                .setOnItemClickListener(new TaskViewHolder.OnItemClickListener() {
                    @Override
                    public void onItemClickListener(TaskViewHolder holder, boolean showDetail) {
                        if (onItemClickListener != null) {
                            onItemClickListener.onItemClickListener(holder, showDetail);
                        }
                    }
                });
    }

    @Override
    public void onBindViewHolder(@NonNull TaskViewHolder holder, int position) {
        if (!cursor.moveToPosition(position)) {
            return;
        }

        holder.data = new TaskDataContainer(cursor);

        holder.binding.checkbox.setChecked(holder.data.isCompleted);
        holder.binding.title.setText(holder.data.title);

        if (holder.data.details.length() < 1) {
            holder.binding.detail.setText("");
            holder.binding.detail.setVisibility(View.GONE);
        } else {
            holder.binding.detail.setText(holder.data.details);
            holder.binding.detail.setVisibility(View.VISIBLE);
        }

        if (holder.data.showCount) {
            long currentTime = System.currentTimeMillis();
            int totalDays = (int) Math.ceil((currentTime - holder.data.taskCreationTime) / 1000.0 / 60 / 60 / 24);
            holder.binding.taskCountStat.setText(context.getString(R.string.task_count_stat, holder.data.count, totalDays));
            holder.binding.taskCountStat.setVisibility(View.VISIBLE);
        } else {
            holder.binding.taskCountStat.setVisibility(View.GONE);
        }
    }

    @Override
    public int getItemCount() {
        return cursor.getCount();
    }


    public void updateCursor(Cursor cursor) {
        this.cursor = cursor;
        notifyDataSetChanged();
    }


    // callbacks
    public interface OnItemClickListener {
        void onItemClickListener(TaskViewHolder holder, boolean showDetail);
    }

    private OnItemClickListener onItemClickListener;

    public TaskRecycleListAdapter setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
        return this;
    }
}

File: app/src/main/java/in/kashewdevelopers/todo/adapters/TaskRecycleListAdapter.kt
package `in`.kashewdevelopers.todo.adapters

import `in`.kashewdevelopers.todo.R
import `in`.kashewdevelopers.todo.data_container.TaskDataContainer
import android.content.Context
import android.database.Cursor
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

class TaskRecycleListAdapter(private var cursor: Cursor, private var context: Context) : RecyclerView.Adapter<TaskViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TaskViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.task_list_item_layout, parent, false)

        return TaskViewHolder(view)
                .setOnItemClickListener(object : TaskViewHolder.OnItemClickListener {
                    override fun onItemClickListener(holder: TaskViewHolder, showDetail: Boolean) {
                        onItemClickListener?.onItemClickListener(holder, showDetail)
                    }
                })
    }

    override fun onBindViewHolder(holder: TaskViewHolder, position: Int) {
        if (!cursor.moveToPosition(position)) {
            return
        }

        holder.data = TaskDataContainer(cursor)

        holder.binding.checkbox.isChecked = holder.data.isCompleted
        holder.binding.title.text = holder.data.title

        holder.binding.detail.text = holder.data.details
        holder.binding.detail.visibility = if (holder.data.details.isEmpty()) View.GONE else View.VISIBLE

        if (holder.data.showCount) {
            holder.binding.taskCountStat.text = context.getString(R.string.task_count_stat, holder.data.count, holder.data.total)
            holder.binding.taskCountStat.visibility = View.VISIBLE
        } else {
            holder.binding.taskCountStat.visibility = View.GONE
        }
    }

    override fun getItemCount(): Int = cursor.count

    fun updateCursor(cursor: Cursor) {
        this.cursor = cursor
        notifyDataSetChanged()
    }


    // callbacks
    interface OnItemClickListener {
        fun onItemClickListener(holder: TaskViewHolder, showDetail: Boolean)
    }

    private var onItemClickListener: OnItemClickListener? = null

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener?): TaskRecycleListAdapter {
        this.onItemClickListener = onItemClickListener
        return this
    }
}
